src/pin_vec.rs

```diff
@@ -23,13 +23,25 @@ impl<T> PinVec<T> {
     }
 
     pub fn clear(&mut self) {
-        for i in 0..self.len {
-            // Safety: we know the pointer is initialized because its index is in bounds, and it
-            // can be dropped because we are emptying the container, which means these contents
-            // will not be accessed again.
-            unsafe { drop_in_place(self.get_mut(i).unwrap()) }
+        if mem::needs_drop::<T>() {
+            let (last_slot, offset, _) = calculate_key(self.len());
+            for (i, mut slot) in self.slots.drain(..).enumerate() {
+                let slice: &mut [MaybeUninit<T>] = if i < last_slot {
+                    &mut *slot
+                } else {
+                    &mut slot[0..offset]
+                };
+                // Safety: we initialized slice to only point to the already-initialized elements.
+                // It's safe to drop_in_place because we are draining the Vec.
+                unsafe {
+                    let slice: &mut [T] = mem::transmute(slice);
```

Comment:
I think I'm partial to casting to a pointer above and reconstituting it as a slice over transmuting a slice. But both are equally correct to me so I don't mind it.

Here's that version:

```rust
let len = if i < last_slot { slot.len() } else { offset };
drop_in_place(from_raw_parts_mut(slot.as_mut_ptr() as *mut T, len));
```